rclcpp/include/rclcpp/rate.hpp

```diff
@@ -112,8 +120,78 @@ class GenericRate : public RateBase
   std::chrono::time_point<Clock, ClockDurationNano> last_interval_;
 };
 
-using Rate = GenericRate<std::chrono::system_clock>;
-using WallRate = GenericRate<std::chrono::steady_clock>;
+class Rate : public RateBase
+{
+public:
+  RCLCPP_SMART_PTR_DEFINITIONS(Rate)
+
+  explicit Rate(
+    const double rate,
+    Clock::SharedPtr clock = std::make_shared<Clock>(RCL_SYSTEM_TIME))
+  : clock_(clock), period_(0, 0), last_interval_(clock_->now())
+  {
+    if (rate <= 0.0) {
+      throw std::invalid_argument{"rate must be greater than 0"};
+    }
+    period_ = Duration::from_seconds(1.0 / rate);
+  }
+  explicit Rate(
+    const Duration & period,
+    Clock::SharedPtr clock = std::make_shared<Clock>(RCL_SYSTEM_TIME))
+  : clock_(clock), period_(period), last_interval_(clock_->now())
+  {
+    if (period <= Duration(0, 0)) {
+      throw std::invalid_argument{"period must be greater than 0"};
+    }
+  }
+
+  virtual bool
+  sleep();
+
+  [[deprecated("use get_type() instead")]]
+  virtual bool
+  is_steady() const;
+
+  virtual rcl_clock_type_t
+  get_type() const;
+
+  virtual void
+  reset();
+
+  Duration
+  period() const;
+
+private:
+  RCLCPP_DISABLE_COPY(Rate)
+
+  Clock::SharedPtr clock_;
+  Duration period_;
+  Time last_interval_;
+};
+
+class WallRate : public Rate
+{
+public:
+  explicit WallRate(const double rate)
+  : Rate(rate, std::make_shared<Clock>(RCL_STEADY_TIME))
+  {}
+
+  explicit WallRate(const Duration & period)
+  : Rate(period, std::make_shared<Clock>(RCL_STEADY_TIME))
+  {}
+};
+
+class ROSRate : public Rate
+{
+public:
+  explicit ROSRate(const double rate)
+  : Rate(rate, std::make_shared<Clock>(RCL_ROS_TIME))
+  {}
+
+  explicit ROSRate(const Duration & period)
+  : Rate(period, std::make_shared<Clock>(RCL_ROS_TIME))
+  {}
```

Comment:
And let's move these constructors to rate.cpp as well (sorry I'm being pedantic here, but I'm currently on a quest to reduce compile times in rclcpp and this is one way to help with that).

Reply:
Yep, no problems. Done in last commit